File: dashboard/server/tests/test_task019_connector_ops.mjs
/**
 * dashboard/server/tests/test_task019_connector_ops.mjs
 * "ConnectorRef reader/model + connector.set/remove ops" (task-019,
 * feature-007-connectors-list, delivery-003, work-017-cli-improvements) --
 * Node twin.
 *
 * server.mjs self-executes on import (parses argv, binds a socket
 * immediately -- no main-guard), so its internal functions (dispatchOp,
 * OP_TABLE, validateConnectorSetArgs, opConnectorSetArgv, ...) are not
 * importable as-is. This file reuses test_task017_registry_tooling_round_
 * trips.py's `_sliced_server_mjs_source` cut point (the same MAIN_MARKER
 * comment, "// Main: parse args, create server, bind, register SIGTERM") to
 * export { dispatchOp, OP_TABLE } from a slice of the REAL source, so every
 * case below drives the ACTUAL production functions -- no reimplementation,
 * no "kept in lockstep by inspection" gap. Connector ops need NO further
 * substitution (unlike task-017's aid-CLI-backed rows): write-connector.sh is
 * dispatched via runWriter, which always spawns bash regardless of host OS
 * (the KI-009 OS-branch lives in runAidCli, never touched here).
 *
 * Covers:
 *   1. OP_TABLE row shape: connector.set / connector.remove carry the
 *      expected scope/writer/argSchema/semanticValidate, no statusMap
 *      override.
 *   2. validateConnectorSetArgs / validateConnectorRemoveArgs: name/type/
 *      endpoint/auth/secret_ref/stem cases (mirrors the Python twin's own
 *      matrix).
 *   3. Full dispatchOp round-trips through the REAL co-vendored
 *      write-connector.sh writer (bash) -- 200 happy paths (descriptor
 *      authored, INDEX.md regenerated), 422 short-circuit before any spawn.
 *
 * LOCAL TEST NOTE: no live server.mjs process is spawned and no port is
 * bound anywhere in this file (the slice is imported in-process via a
 * file:// URL) -- every case here calls dispatchOp directly, the same "no
 * server spawn" convention test_task004_op_dispatch.py's real-writer
 * round-trips already established -- safe to run locally per the project's
 * port-binding-server-test constraint, and exercised directly as part of
 * this task's own verification pass.
 *
 * Run: node dashboard/server/tests/test_task019_connector_ops.mjs
 *
 * ASCII-only source. Node built-in modules only.
 */

import { readFileSync, mkdirSync, writeFileSync, rmSync, existsSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath, pathToFileURL } from "url";
import { tmpdir } from "os";
import { randomBytes } from "crypto";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const SERVER_MJS_PATH = join(__dirname, "..", "server.mjs");
const MAIN_MARKER = "// Main: parse args, create server, bind, register SIGTERM";

let passed = 0;
let failed = 0;

function assert(cond, msg) {
  if (cond) {
    process.stdout.write("  PASS: " + msg + "\n");
    passed++;
  } else {
    process.stderr.write("  FAIL: " + msg + "\n");
    failed++;
  }
}

function assertEquals(a, b, msg) {
  assert(a === b, msg + " -- got " + JSON.stringify(a) + " expected " + JSON.stringify(b));
}

// ---------------------------------------------------------------------------
// Slice server.mjs before the self-executing 'Main' tail (mirrors
// test_task017_registry_tooling_round_trips.py's _sliced_server_mjs_source(),
// minus the AID_CLI_PATH/timeout/KI-009 substitutions that task's aid-CLI-
// backed rows need -- connector ops never touch those constants).
// ---------------------------------------------------------------------------
function sliceServerSource() {
  const text = readFileSync(SERVER_MJS_PATH, "utf8");
  const idx = text.indexOf(MAIN_MARKER);
  if (idx === -1) {
    throw new Error(
      "server.mjs 'Main' marker is gone -- this test's slice cut point needs updating"
    );
  }
  return text.slice(0, idx) + "\nexport { dispatchOp, OP_TABLE };\n";
}

// Written INTO dashboard/server/ (colocated with reader.mjs), NOT the OS tmp
// dir -- server.mjs's own `import ... from "./reader.mjs"` is relative, so the
// slice must sit beside reader.mjs to resolve (mirrors test_task017's own
// _SERVER_DIR placement for its sliced copy).
const sliceDir = join(__dirname, "..");
const slicePath = join(sliceDir, `_test_task019_slice_${randomBytes(8).toString("hex")}.mjs`);
writeFileSync(slicePath, sliceServerSource(), "utf8");

let dispatchOp;
let OP_TABLE;
try {
  const mod = await import(pathToFileURL(slicePath).href);
  dispatchOp = mod.dispatchOp;
  OP_TABLE = mod.OP_TABLE;
} finally {
  rmSync(slicePath, { force: true });
}

// ---------------------------------------------------------------------------
// Scratch repo helper
// ---------------------------------------------------------------------------

function makeTmpRepo() {
  const root = join(tmpdir(), `_test_task019_repo_${randomBytes(8).toString("hex")}`);
  mkdirSync(root, { recursive: true });
  return root;
}

function cleanup(root) {
  rmSync(root, { recursive: true, force: true });
}

// ===========================================================================
// (1) OP_TABLE row shape
// ===========================================================================

function testOpTableRowShape() {
  const setRow = OP_TABLE["connector.set"];
  assertEquals(setRow.scope, "project", "connector.set scope is 'project'");
  assertEquals(setRow.writer, "write-connector.sh", "connector.set writer is write-connector.sh");
  assert(setRow.argSchema.name.required === true, "connector.set argSchema.name required");
  assert(setRow.argSchema.type.required === true, "connector.set argSchema.type required");
  for (const key of ["endpoint", "auth", "secret_ref"]) {
    assert(Object.prototype.hasOwnProperty.call(setRow.argSchema, key), `connector.set argSchema has ${key}`);
    assert(!setRow.argSchema[key].required, `connector.set argSchema.${key} not required`);
  }
  assert(typeof setRow.buildArgv === "function", "connector.set buildArgv is a function");
  assert(typeof setRow.semanticValidate === "function", "connector.set semanticValidate is a function");
  assert(setRow.statusMap === null, "connector.set statusMap is null (uses DEFAULT_MAP)");

  const removeRow = OP_TABLE["connector.remove"];
  assertEquals(removeRow.scope, "project", "connector.remove scope is 'project'");
  assertEquals(removeRow.writer, "write-connector.sh", "connector.remove writer is write-connector.sh");
  assert(removeRow.argSchema.stem.required === true, "connector.remove argSchema.stem required");
  assert(typeof removeRow.buildArgv === "function", "connector.remove buildArgv is a function");
  assert(typeof removeRow.semanticValidate === "function", "connector.remove semanticValidate is a function");
  assert(removeRow.statusMap === null, "connector.remove statusMap is null (uses DEFAULT_MAP)");
}

// ===========================================================================
// (2) Pure semantic validation (via OP_TABLE's own semanticValidate hook)
// ===========================================================================

function validateSet(args) {
  return OP_TABLE["connector.set"].semanticValidate(args);
}

function validateRemove(args) {
  return OP_TABLE["connector.remove"].semanticValidate(args);
}

function testValidateConnectorSetArgs() {
  assertEquals(validateSet({ name: "GitHub", type: "mcp" }), null, "valid mcp name/type passes");
  assert(validateSet({ name: "", type: "mcp" }) !== null, "empty name rejected");
  assert(validateSet({ name: "x".repeat(81), type: "mcp" }) !== null, "overlong name rejected");
  assertEquals(validateSet({ name: "x".repeat(80), type: "mcp" }), null, "name at max length passes");
  assert(validateSet({ name: "a\nb", type: "mcp" }) !== null, "name with newline rejected");
  assert(validateSet({ name: "a|b", type: "mcp" }) !== null, "name with pipe rejected");
  assert(validateSet({ name: "a\x01b", type: "mcp" }) !== null, "name with control char rejected");

  assert(validateSet({ name: "n", type: "ftp" }) !== null, "invalid type rejected");
  assertEquals(
    validateSet({ name: "n", type: "api", endpoint: "https://x", auth: "token" }),
    null, "valid api type with required fields passes"
  );
  assertEquals(
    validateSet({ name: "n", type: "ssh", endpoint: "host:22" }),
    null, "valid ssh type (auth omitted, writer forces ssh-key) passes"
  );

  for (const ctype of ["api", "ssh", "url", "cli"]) {
    const args = { name: "n", type: ctype };
    if (ctype !== "ssh") args.auth = "token";
    assert(validateSet(args) !== null, `endpoint required for type ${ctype}`);
  }
  assertEquals(validateSet({ name: "n", type: "mcp" }), null, "endpoint optional for mcp");

  for (const ctype of ["api", "url", "cli"]) {
    assert(
      validateSet({ name: "n", type: ctype, endpoint: "https://x" }) !== null,
      `auth required for type ${ctype}`
    );
  }
  assert(
    validateSet({ name: "n", type: "api", endpoint: "https://x", auth: "bearer" }) !== null,
    "invalid auth enum rejected"
  );
  assert(
    validateSet({ name: "n", type: "api", endpoint: "x".repeat(201), auth: "token" }) !== null,
    "overlong endpoint rejected"
  );
  assert(
    validateSet({ name: "n", type: "api", endpoint: "https://x|evil", auth: "token" }) !== null,
    "endpoint with pipe rejected"
  );

  assertEquals(
    validateSet({ name: "n", type: "api", endpoint: "https://x", auth: "token" }),
    null, "omitted secret_ref never rejected"
  );
  assertEquals(
    validateSet({
      name: "n", type: "api", endpoint: "https://x", auth: "token", secret_ref: "env:MY_TOKEN",
    }),
    null, "env: form accepted"
  );
  assertEquals(
    validateSet({ name: "n", type: "ssh", endpoint: "host", secret_ref: "file:.aid/connectors/.secrets/n" }),
    null, "file: form accepted"
  );
  assert(
    validateSet({
      name: "n", type: "api", endpoint: "https://x", auth: "token", secret_ref: "not-a-valid-ref",
    }) !== null,
    "malformed secret_ref rejected"
  );
  assert(
    validateSet({ name: "n", type: "mcp", secret_ref: "env:MY_TOKEN" }) !== null,
    "secret_ref forbidden for mcp"
  );
  assert(
    validateSet({
      name: "n", type: "url", endpoint: "https://x", auth: "none", secret_ref: "env:MY_TOKEN",
    }) !== null,
    "secret_ref forbidden for auth none"
  );
}

function testValidateConnectorRemoveArgs() {
  assertEquals(validateRemove({ stem: "github" }), null, "valid stem passes");
  assertEquals(validateRemove({ stem: "my-connector-2" }), null, "valid stem with digits/dashes passes");
  assert(validateRemove({ stem: "GitHub" }) !== null, "uppercase stem rejected");
  assert(validateRemove({ stem: "-github" }) !== null, "stem starting with dash rejected");
  assert(validateRemove({ stem: "a/b" }) !== null, "stem with slash rejected");
  assert(validateRemove({ stem: "../etc" }) !== null, "stem with dotdot rejected");
}

// ===========================================================================
// (3) Full dispatchOp round-trips through the REAL write-connector.sh writer
// ===========================================================================

function readFile(path) {
  return readFileSync(path, "utf8");
}

async function testConnectorSetMcpSuccess() {
  const root = makeTmpRepo();
  try {
    const [status, body] = dispatchOp(
      OP_TABLE, { op: "connector.set", target: {}, args: { name: "GitHub", type: "mcp" } }, root
    );
    assertEquals(status, 200, "connector.set mcp success status");
    const parsed = JSON.parse(Buffer.from(body).toString("utf-8"));
    assertEquals(JSON.stringify(parsed), JSON.stringify({ ok: true, op: "connector.set" }), "connector.set ok envelope");
    const descriptorPath = join(root, ".aid", "connectors", "github.md");
    assert(existsSync(descriptorPath), "github.md descriptor written");
    const text = readFile(descriptorPath);
    assert(text.includes("connection_type: mcp"), "descriptor carries connection_type: mcp");
    assert(text.includes("auth_method: none"), "descriptor carries auth_method: none");
    assert(!text.includes("secret_reference:"), "descriptor carries no secret_reference for mcp");
    assert(existsSync(join(root, ".aid", "connectors", "INDEX.md")), "INDEX.md regenerated");
  } finally {
    cleanup(root);
  }
}

async function testConnectorSetApiWithDefaultSecretRef() {
  const root = makeTmpRepo();
  try {
    const [status] = dispatchOp(
      OP_TABLE,
      {
        op: "connector.set", target: {},
        args: { name: "Jira", type: "api", endpoint: "https://acme.atlassian.net/rest/api/3", auth: "token" },
      },
      root
    );
    assertEquals(status, 200, "connector.set api success status");
    const text = readFile(join(root, ".aid", "connectors", "jira.md"));
    assert(text.includes("auth_method: token"), "descriptor carries auth_method: token");
    assert(
      text.includes('secret_reference: "file:.aid/connectors/.secrets/jira"'),
      "descriptor carries the default secret_reference form"
    );
  } finally {
    cleanup(root);
  }
}

async function testConnectorSetSemanticFailureIs422BeforeSpawn() {
  const root = makeTmpRepo();
  try {
    const [status, body] = dispatchOp(
      OP_TABLE, { op: "connector.set", target: {}, args: { name: "n", type: "ftp" } }, root
    );
    assertEquals(status, 422, "invalid type maps to 422");
    const parsed = JSON.parse(Buffer.from(body).toString("utf-8"));
    assertEquals(parsed.error, "invalid-value", "error class is invalid-value");
    assert(!existsSync(join(root, ".aid", "connectors")), "no spawn happened (connectors dir never created)");
  } finally {
    cleanup(root);
  }
}

async function testConnectorRemoveSuccess() {
  const root = makeTmpRepo();
  try {
    const [setStatus] = dispatchOp(
      OP_TABLE, { op: "connector.set", target: {}, args: { name: "GitHub", type: "mcp" } }, root
    );
    assertEquals(setStatus, 200, "seed connector.set succeeds");
    const descriptorPath = join(root, ".aid", "connectors", "github.md");
    assert(existsSync(descriptorPath), "seeded descriptor exists before remove");

    const [status, body] = dispatchOp(
      OP_TABLE, { op: "connector.remove", target: {}, args: { stem: "github" } }, root
    );
    assertEquals(status, 200, "connector.remove success status");
    const parsed = JSON.parse(Buffer.from(body).toString("utf-8"));
    assertEquals(JSON.stringify(parsed), JSON.stringify({ ok: true, op: "connector.remove" }), "connector.remove ok envelope");
    assert(!existsSync(descriptorPath), "descriptor removed from disk");
  } finally {
    cleanup(root);
  }
}

async function testConnectorRemoveBadStemIs422BeforeSpawn() {
  const root = makeTmpRepo();
  try {
    const [status, body] = dispatchOp(
      OP_TABLE, { op: "connector.remove", target: {}, args: { stem: "../etc" } }, root
    );
    assertEquals(status, 422, "bad stem maps to 422");
    const parsed = JSON.parse(Buffer.from(body).toString("utf-8"));
    assertEquals(parsed.error, "invalid-value", "error class is invalid-value");
  } finally {
    cleanup(root);
  }
}

async function testTargetWorkIdIgnoredForProjectScope() {
  const root = makeTmpRepo();
  try {
    const [status] = dispatchOp(
      OP_TABLE,
      { op: "connector.set", target: { work_id: "work-999-does-not-exist" }, args: { name: "GitHub", type: "mcp" } },
      root
    );
    assertEquals(status, 200, "project-scoped op ignores target.work_id (never 404s)");
  } finally {
    cleanup(root);
  }
}

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------

testOpTableRowShape();
testValidateConnectorSetArgs();
testValidateConnectorRemoveArgs();
await testConnectorSetMcpSuccess();
await testConnectorSetApiWithDefaultSecretRef();
await testConnectorSetSemanticFailureIs422BeforeSpawn();
await testConnectorRemoveSuccess();
await testConnectorRemoveBadStemIs422BeforeSpawn();
await testTargetWorkIdIgnoredForProjectScope();

process.stdout.write(`\n${passed} passed, ${failed} failed\n`);
process.exit(failed > 0 ? 1 : 0);

File: dashboard/server/tests/test_task019_connector_ops.py
"""
test_task019_connector_ops.py -- "ConnectorRef reader/model + connector.set/remove
ops" (task-019, feature-007-connectors-list, delivery-003,
work-017-cli-improvements) -- Python twin.

Covers, all in-process (no socket bind -- see LOCAL TEST NOTE below):

  1. OP_TABLE shape: connector.set / connector.remove rows carry the expected
     scope ("project" -- no work_id), writer ("write-connector.sh"), arg_schema
     keys, and semantic_validate hooks; no per-op status_map override (relies
     on feature-001's generic DEFAULT_MAP, which already covers the writer's
     0/4/5/3 exit alphabet).
  2. Pure semantic validation (_validate_connector_set_args /
     _validate_connector_remove_args): name length/charset, type enum,
     conditional endpoint/auth requiredness per type, secret_ref pattern +
     forbidden-for-mcp/auth-none, stem charset.
  3. Argv builders (_op_connector_set_argv / _op_connector_remove_argv): argv
     is an array (never a shell string), starts with the write-connector.sh
     subcommand, --root is server-built from served_root (never the request
     body), optional flags omitted when absent.
  4. Full _dispatch_op round-trips through the REAL co-vendored
     write-connector.sh writer (bash) -- 200 happy paths (descriptor authored,
     INDEX.md regenerated) for connector.set/connector.remove, and 422
     'invalid-value' short-circuits BEFORE any child spawn for a semantically
     invalid request (proven via a served_root with no .aid/connectors/ at
     all -- if the writer had been invoked, a missing root would still exit 0
     since write-connector.sh mkdir -p's it, so a 422 here is FIRM proof the
     pre-validation fired before spawn).
  5. target: {} (present-but-empty) request shape -- no target.work_id is ever
     consumed by either op (project-scoped, feature-002 settings.set envelope
     precedent).

LOCAL TEST NOTE (Windows/Git-Bash host, no setsid): every class in this file
calls `srv._dispatch_op(...)` directly -- no `_ServerThread` socket bind
anywhere -- so the whole file is safe to run locally per the project's
port-binding-server-test constraint. All classes were exercised directly as
part of this task's own verification pass.

Python 3.11+ stdlib only. No third-party deps.
"""

from __future__ import annotations

import json
import shutil
import sys
import tempfile
import unittest
from pathlib import Path

# ---------------------------------------------------------------------------
# Make the dashboard package importable regardless of CWD (mirrors the other
# task-0NN suites' own sys.path setup).
# ---------------------------------------------------------------------------
_TESTS_DIR = Path(__file__).resolve().parent         # dashboard/server/tests/
_SERVER_DIR = _TESTS_DIR.parent                      # dashboard/server/
_DASHBOARD_DIR = _SERVER_DIR.parent                  # dashboard/
_REPO_ROOT = _DASHBOARD_DIR.parent                   # AID/

if str(_REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(_REPO_ROOT))

from dashboard.server import server as srv


class _TmpRepo:
    """Context manager: a scratch repo root, cleaned up on exit."""

    def __enter__(self) -> Path:
        self.path = Path(tempfile.mkdtemp())
        return self.path

    def __exit__(self, *_exc) -> None:
        shutil.rmtree(str(self.path), ignore_errors=True)


# ===========================================================================
# (1) OP_TABLE shape
# ===========================================================================

class TestOpTableConnectorRows(unittest.TestCase):
    def test_connector_set_row_shape(self):
        row = srv.OP_TABLE["connector.set"]
        self.assertEqual(row["scope"], "project")
        self.assertEqual(row["writer"], "write-connector.sh")
        self.assertIn("name", row["arg_schema"])
        self.assertTrue(row["arg_schema"]["name"]["required"])
        self.assertIn("type", row["arg_schema"])
        self.assertTrue(row["arg_schema"]["type"]["required"])
        for optional_key in ("endpoint", "auth", "secret_ref"):
            self.assertIn(optional_key, row["arg_schema"])
            self.assertNotIn("required", row["arg_schema"][optional_key])
        self.assertTrue(callable(row["build_argv"]))
        self.assertTrue(callable(row["semantic_validate"]))
        self.assertIsNone(row.get("status_map"))

    def test_connector_remove_row_shape(self):
        row = srv.OP_TABLE["connector.remove"]
        self.assertEqual(row["scope"], "project")
        self.assertEqual(row["writer"], "write-connector.sh")
        self.assertIn("stem", row["arg_schema"])
        self.assertTrue(row["arg_schema"]["stem"]["required"])
        self.assertTrue(callable(row["build_argv"]))
        self.assertTrue(callable(row["semantic_validate"]))
        self.assertIsNone(row.get("status_map"))


# ===========================================================================
# (2) Pure semantic validation
# ===========================================================================

class TestValidateConnectorSetArgsName(unittest.TestCase):
    def test_valid_name_passes_for_mcp(self):
        self.assertIsNone(srv._validate_connector_set_args({"name": "GitHub", "type": "mcp"}))

    def test_empty_name_rejected(self):
        err = srv._validate_connector_set_args({"name": "", "type": "mcp"})
        self.assertIsNotNone(err)

    def test_overlong_name_rejected(self):
        err = srv._validate_connector_set_args({"name": "x" * 81, "type": "mcp"})
        self.assertIsNotNone(err)

    def test_name_at_max_length_passes(self):
        self.assertIsNone(srv._validate_connector_set_args({"name": "x" * 80, "type": "mcp"}))

    def test_name_with_newline_rejected(self):
        err = srv._validate_connector_set_args({"name": "a\nb", "type": "mcp"})
        self.assertIsNotNone(err)

    def test_name_with_pipe_rejected(self):
        err = srv._validate_connector_set_args({"name": "a|b", "type": "mcp"})
        self.assertIsNotNone(err)

    def test_name_with_control_char_rejected(self):
        err = srv._validate_connector_set_args({"name": "a\x01b", "type": "mcp"})
        self.assertIsNotNone(err)


class TestValidateConnectorSetArgsType(unittest.TestCase):
    def test_valid_types_pass_with_required_fields(self):
        cases = [
            {"name": "n", "type": "mcp"},
            {"name": "n", "type": "api", "endpoint": "https://x", "auth": "token"},
            {"name": "n", "type": "ssh", "endpoint": "host:22"},
            {"name": "n", "type": "url", "endpoint": "https://x", "auth": "none"},
            {"name": "n", "type": "cli", "endpoint": "cmd", "auth": "pat"},
        ]
        for args in cases:
            with self.subTest(args=args):
                self.assertIsNone(srv._validate_connector_set_args(args))

    def test_invalid_type_rejected(self):
        err = srv._validate_connector_set_args({"name": "n", "type": "ftp"})
        self.assertIsNotNone(err)
        self.assertIn("type", err)


class TestValidateConnectorSetArgsEndpointAuth(unittest.TestCase):
    def test_endpoint_required_for_aid_managed_types(self):
        for ctype in ("api", "ssh", "url", "cli"):
            with self.subTest(ctype=ctype):
                args = {"name": "n", "type": ctype}
                if ctype in ("api", "url", "cli"):
                    args["auth"] = "token"
                err = srv._validate_connector_set_args(args)
                self.assertIsNotNone(err, f"endpoint should be required for {ctype}")

    def test_endpoint_optional_for_mcp(self):
        self.assertIsNone(srv._validate_connector_set_args({"name": "n", "type": "mcp"}))
        self.assertIsNone(
            srv._validate_connector_set_args({"name": "n", "type": "mcp", "endpoint": "info"})
        )

    def test_auth_required_for_api_url_cli(self):
        for ctype in ("api", "url", "cli"):
            with self.subTest(ctype=ctype):
                err = srv._validate_connector_set_args(
                    {"name": "n", "type": ctype, "endpoint": "https://x"}
                )
                self.assertIsNotNone(err)

    def test_auth_optional_for_ssh_and_mcp(self):
        self.assertIsNone(
            srv._validate_connector_set_args({"name": "n", "type": "ssh", "endpoint": "host"})
        )
        self.assertIsNone(srv._validate_connector_set_args({"name": "n", "type": "mcp"}))

    def test_invalid_auth_enum_rejected(self):
        err = srv._validate_connector_set_args(
            {"name": "n", "type": "api", "endpoint": "https://x", "auth": "bearer"}
        )
        self.assertIsNotNone(err)

    def test_endpoint_over_length_rejected(self):
        err = srv._validate_connector_set_args(
            {"name": "n", "type": "api", "endpoint": "x" * 201, "auth": "token"}
        )
        self.assertIsNotNone(err)

    def test_endpoint_with_pipe_rejected(self):
        err = srv._validate_connector_set_args(
            {"name": "n", "type": "api", "endpoint": "https://x|evil", "auth": "token"}
        )
        self.assertIsNotNone(err)


class TestValidateConnectorSetArgsSecretRef(unittest.TestCase):
    def test_omitted_secret_ref_never_rejected(self):
        self.assertIsNone(
            srv._validate_connector_set_args(
                {"name": "n", "type": "api", "endpoint": "https://x", "auth": "token"}
            )
        )

    def test_env_form_accepted(self):
        self.assertIsNone(
            srv._validate_connector_set_args({
                "name": "n", "type": "api", "endpoint": "https://x", "auth": "token",
                "secret_ref": "env:MY_TOKEN",
            })
        )

    def test_file_form_accepted(self):
        self.assertIsNone(
            srv._validate_connector_set_args({
                "name": "n", "type": "ssh", "endpoint": "host",
                "secret_ref": "file:.aid/connectors/.secrets/n",
            })
        )

    def test_keychain_form_accepted(self):
        self.assertIsNone(
            srv._validate_connector_set_args({
                "name": "n", "type": "cli", "endpoint": "cmd", "auth": "pat",
                "secret_ref": "keychain:my-key",
            })
        )

    def test_malformed_secret_ref_rejected(self):
        err = srv._validate_connector_set_args({
            "name": "n", "type": "api", "endpoint": "https://x", "auth": "token",
            "secret_ref": "not-a-valid-ref",
        })
        self.assertIsNotNone(err)

    def test_env_form_bad_var_name_rejected(self):
        err = srv._validate_connector_set_args({
            "name": "n", "type": "api", "endpoint": "https://x", "auth": "token",
            "secret_ref": "env:123bad",
        })
        self.assertIsNotNone(err)

    def test_secret_ref_forbidden_for_mcp(self):
        err = srv._validate_connector_set_args({
            "name": "n", "type": "mcp", "secret_ref": "env:MY_TOKEN",
        })
        self.assertIsNotNone(err)

    def test_secret_ref_forbidden_for_auth_none(self):
        err = srv._validate_connector_set_args({
            "name": "n", "type": "url", "endpoint": "https://x", "auth": "none",
            "secret_ref": "env:MY_TOKEN",
        })
        self.assertIsNotNone(err)

    def test_secret_ref_over_length_rejected(self):
        err = srv._validate_connector_set_args({
            "name": "n", "type": "api", "endpoint": "https://x", "auth": "token",
            "secret_ref": "file:" + ("x" * 200),
        })
        self.assertIsNotNone(err)


class TestValidateConnectorRemoveArgs(unittest.TestCase):
    def test_valid_stem_passes(self):
        self.assertIsNone(srv._validate_connector_remove_args({"stem": "github"}))
        self.assertIsNone(srv._validate_connector_remove_args({"stem": "my-connector-2"}))

    def test_uppercase_stem_rejected(self):
        self.assertIsNotNone(srv._validate_connector_remove_args({"stem": "GitHub"}))

    def test_stem_starting_with_dash_rejected(self):
        self.assertIsNotNone(srv._validate_connector_remove_args({"stem": "-github"}))

    def test_stem_with_slash_rejected(self):
        self.assertIsNotNone(srv._validate_connector_remove_args({"stem": "a/b"}))

    def test_stem_with_dotdot_rejected(self):
        self.assertIsNotNone(srv._validate_connector_remove_args({"stem": "../etc"}))


# ===========================================================================
# (3) Argv builders
# ===========================================================================

class TestConnectorSetArgvBuilder(unittest.TestCase):
    def test_argv_is_list_and_starts_with_set_subcommand(self):
        argv, env = srv._op_connector_set_argv(
            None, "/repo/root", {}, {"name": "Jira", "type": "api", "endpoint": "https://x", "auth": "token"},
        )
        self.assertIsInstance(argv, list)
        self.assertEqual(argv[0], "set")
        self.assertIn("--root", argv)
        root_idx = argv.index("--root")
        self.assertTrue(argv[root_idx + 1].endswith(".aid/connectors"))

    def test_optional_flags_omitted_when_absent(self):
        argv, _ = srv._op_connector_set_argv(None, "/repo/root", {}, {"name": "GitHub", "type": "mcp"})
        self.assertNotIn("--endpoint", argv)
        self.assertNotIn("--auth", argv)
        self.assertNotIn("--secret-ref", argv)

    def test_optional_flags_included_when_present(self):
        argv, _ = srv._op_connector_set_argv(
            None, "/repo/root", {},
            {"name": "Jira", "type": "api", "endpoint": "https://x", "auth": "token", "secret_ref": "env:X"},
        )
        self.assertIn("--endpoint", argv)
        self.assertIn("--auth", argv)
        self.assertIn("--secret-ref", argv)

    def test_root_never_taken_from_args_body(self):
        """--root is ALWAYS server-built from served_root, never echoed from a
        client-supplied field (SEC-2)."""
        argv, _ = srv._op_connector_set_argv(
            None, "/repo/root", {}, {"name": "n", "type": "mcp", "root": "/evil/path"},
        )
        root_idx = argv.index("--root")
        self.assertNotIn("/evil/path", argv[root_idx + 1])


class TestConnectorRemoveArgvBuilder(unittest.TestCase):
    def test_argv_is_list_and_starts_with_remove_subcommand(self):
        argv, env = srv._op_connector_remove_argv(None, "/repo/root", {}, {"stem": "github"})
        self.assertIsInstance(argv, list)
        self.assertEqual(argv[0], "remove")
        self.assertIn("--root", argv)
        self.assertIn("--stem", argv)
        self.assertEqual(argv[argv.index("--stem") + 1], "github")


# ===========================================================================
# (4) Full _dispatch_op round-trips through the REAL write-connector.sh writer
# ===========================================================================

class TestConnectorOpsRealWriterRoundTrips(unittest.TestCase):
    def test_connector_set_mcp_success(self):
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.set", "target": {}, "args": {"name": "GitHub", "type": "mcp"}},
                str(root),
            )
            self.assertEqual(status, 200)
            self.assertEqual(json.loads(body), {"ok": True, "op": "connector.set"})
            descriptor = root / ".aid" / "connectors" / "github.md"
            self.assertTrue(descriptor.is_file())
            text = descriptor.read_text(encoding="utf-8")
            self.assertIn("connection_type: mcp", text)
            self.assertIn("auth_method: none", text)
            self.assertNotIn("secret_reference:", text)
            index_md = root / ".aid" / "connectors" / "INDEX.md"
            self.assertTrue(index_md.is_file())

    def test_connector_set_api_with_default_secret_ref(self):
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {
                    "op": "connector.set", "target": {},
                    "args": {"name": "Jira", "type": "api",
                             "endpoint": "https://acme.atlassian.net/rest/api/3", "auth": "token"},
                },
                str(root),
            )
            self.assertEqual(status, 200)
            descriptor = root / ".aid" / "connectors" / "jira.md"
            text = descriptor.read_text(encoding="utf-8")
            self.assertIn("auth_method: token", text)
            self.assertIn('secret_reference: "file:.aid/connectors/.secrets/jira"', text)

    def test_connector_set_semantic_failure_maps_to_422_before_spawn(self):
        """An invalid --type never reaches the writer (422 'invalid-value')."""
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.set", "target": {}, "args": {"name": "n", "type": "ftp"}},
                str(root),
            )
            self.assertEqual(status, 422)
            self.assertEqual(json.loads(body)["error"], "invalid-value")
            # Proof no spawn happened: write-connector.sh would have mkdir -p'd
            # the connectors/ dir even on failure paths that reach it.
            self.assertFalse((root / ".aid" / "connectors").exists())

    def test_connector_remove_success(self):
        with _TmpRepo() as root:
            # Seed a connector first.
            status, _ = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.set", "target": {}, "args": {"name": "GitHub", "type": "mcp"}},
                str(root),
            )
            self.assertEqual(status, 200)
            descriptor = root / ".aid" / "connectors" / "github.md"
            self.assertTrue(descriptor.is_file())

            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.remove", "target": {}, "args": {"stem": "github"}},
                str(root),
            )
            self.assertEqual(status, 200)
            self.assertEqual(json.loads(body), {"ok": True, "op": "connector.remove"})
            self.assertFalse(descriptor.exists())

    def test_connector_remove_idempotent_for_absent_stem(self):
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.remove", "target": {}, "args": {"stem": "never-existed"}},
                str(root),
            )
            self.assertEqual(status, 200)

    def test_connector_remove_bad_stem_422_before_spawn(self):
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.remove", "target": {}, "args": {"stem": "../etc"}},
                str(root),
            )
            self.assertEqual(status, 422)
            self.assertEqual(json.loads(body)["error"], "invalid-value")


# ===========================================================================
# (5) target: {} envelope shape -- no work_id consumed
# ===========================================================================

class TestConnectorOpsProjectScoped(unittest.TestCase):
    def test_missing_target_key_defaults_to_empty_and_still_dispatches(self):
        """The op does not require 'target' in the body at all (defaults to {})."""
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {"op": "connector.set", "args": {"name": "GitHub", "type": "mcp"}},
                str(root),
            )
            self.assertEqual(status, 200)

    def test_target_with_work_id_is_ignored_not_consumed(self):
        """scope='project' never triggers work_id resolution (unlike 'task'/'pipeline')."""
        with _TmpRepo() as root:
            status, body = srv._dispatch_op(
                srv.OP_TABLE,
                {
                    "op": "connector.set",
                    "target": {"work_id": "work-999-does-not-exist"},
                    "args": {"name": "GitHub", "type": "mcp"},
                },
                str(root),
            )
            # Must NOT 404 (which would mean work_id resolution was attempted).
            self.assertEqual(status, 200)


if __name__ == "__main__":
    unittest.main()
